soc/bouffalolab/bl60x/soc.c

```diff
@@ -0,0 +1,890 @@
+
+	 /* Turn PLL back ON */
+	 /* frequency stabilization ? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_PU_CLKPLL_SFREG_UMSK)
+	    | ((uint32_t)(1) << PDS_PU_CLKPLL_SFREG_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	 /* let settle for a while (5 us in SDK), we may not be
+	  * running at 32Mhz right now
+	  */
+	system_clock_delay_32M_ms(2);
+
+	 /* enable PPL clock actual? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_PU_CLKPLL_UMSK)
+	    | ((uint32_t)(1) << PDS_PU_CLKPLL_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	/* More power up sequencing*/
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_PU_CP_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_CP_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_PFD_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_PFD_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_FBDV_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_FBDV_POS);
+	tmp = (tmp & PDS_CLKPLL_PU_POSTDIV_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_PU_POSTDIV_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	system_clock_delay_32M_ms(2);
+
+	/* reset couple things one by one? */
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDM_RESET_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_SDM_RESET_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_RESET_FBDV_UMSK)
+	    | ((uint32_t)(1) << PDS_CLKPLL_RESET_FBDV_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_RESET_FBDV_UMSK)
+	    | ((uint32_t)(0) << PDS_CLKPLL_RESET_FBDV_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+
+	tmp = sys_read32(PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDM_RESET_UMSK)
+	    | ((uint32_t)(0) << PDS_CLKPLL_SDM_RESET_POS);
+	sys_write32(tmp, PDS_BASE + PDS_PU_RST_CLKPLL_OFFSET);
+}
+
+/* copied over from driver for convenience */
+static uint32_t system_uart_bflb_get_crystal_frequency(void)
+{
+	uint32_t tmp;
+
+	/* get clkpll_sdmin */
+	tmp = sys_read32(PDS_BASE + PDS_CLKPLL_SDM_OFFSET);
+	tmp = (tmp & PDS_CLKPLL_SDMIN_MSK) >> PDS_CLKPLL_SDMIN_POS;
+
+	/*
+	 * The 'Sigma Delta Modulator Input' scales the PLL output relative to the
+	 * input frequency (eg value is x2, output frequency doubles), it's a
+	 * direct indicator of the XCLK (crystal or 32M osc) frequency to go in the
+	 * PLL.
+	 *
+	 * This value can be read from efuse or from the firmware header which will
+	 * be implemented in the future.
+	 */
+	switch (tmp) {
+	case 0x500000:
+		return MHZ(24);
+	case 0x3C0000:
+		return MHZ(32);
+	case 0x320000:
+		return MHZ(38.4);
+	case 0x300000:
+		return MHZ(40);
+	case 0x49D39D:
+		return MHZ(26);
+	default:
+		return MHZ(32);
+	}
+}
```

Comment:
I am not sure what would be the best way to do this, but it sounds like it might be possible to factorize this function with [above](https://github.com/zephyrproject-rtos/zephyr/pull/78795/files#diff-8819552a2c19cb06392c676b593a020af32ed02cd187a491f7dd9a7867307f1cR72).

Maybe this would be the role for clock-control? Maybe there was already some future refactoring planned after this initial support? A comment in the code explaining would save you from having the question over and over.

Reply:
All the clocks will be reworked based on #72102